File: pe-core/src/main/java/org/saar/core/model/InstancedElementsMesh.java
package org.saar.core.model;

import org.saar.lwjgl.opengl.constants.DataType;
import org.saar.lwjgl.opengl.constants.RenderMode;
import org.saar.lwjgl.opengl.objects.vaos.IVao;
import org.saar.lwjgl.opengl.utils.GlRendering;

public class InstancedElementsMesh extends MeshBase implements Mesh {

    private final RenderMode mode;
    private final int count;
    private final DataType indexType;
    private final int instances;

    public InstancedElementsMesh(IVao vao, RenderMode mode, int count, DataType indexType, int instances) {
        super(vao);
        this.mode = mode;
        this.count = count;
        this.indexType = indexType;
        this.instances = instances;
    }

    @Override
    protected void doDrawCall() {
        GlRendering.drawElementsInstanced(this.mode, this.count, this.indexType, 0, this.instances);
    }
}

File: pe-core/src/main/java/org/saar/core/mesh/InstancedElementsMesh.kt
package org.saar.core.mesh

import org.saar.lwjgl.opengl.constants.DataType
import org.saar.lwjgl.opengl.constants.RenderMode
import org.saar.lwjgl.opengl.drawcall.InstancedElementsDrawCall
import org.saar.lwjgl.opengl.vao.IVao

class InstancedElementsMesh(
    private val vao: IVao,
    var drawCall: InstancedElementsDrawCall
) : Mesh {

    fun set(
        renderMode: RenderMode = this.drawCall.renderMode,
        count: Int = this.drawCall.count,
        type: DataType = this.drawCall.type,
        indices: Long = this.drawCall.indices,
        instances: Int = this.drawCall.instances,
    ) {
        this.drawCall = InstancedElementsDrawCall(renderMode, count, type, indices, instances)
    }

    override fun draw() {
        this.vao.bind()
        this.vao.enableAttributes()
        this.drawCall.doDrawCall()
    }

    override fun delete() {
        this.vao.delete()
    }
}